crda/remotematcher.go

```diff
@@ -0,0 +1,389 @@
+package crda
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/quay/zlog"
+	"go.opentelemetry.io/otel/baggage"
+	"go.opentelemetry.io/otel/label"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/libvuln/driver"
+)
+
+var (
+	_ driver.Matcher             = (*Matcher)(nil)
+	_ driver.RemoteMatcher       = (*Matcher)(nil)
+	_ driver.MatcherConfigurable = (*Matcher)(nil)
+)
+
+const (
+	// Bounded concurrency limit.
+	defaultBatchSize          = 10
+	defaultEndPoint           = "/api/v2/vulnerability-analysis"
+	defaultRequestConcurrency = 10
+	defaultURL                = "https://f8a-analytics-2445582058137.production.gw.apicast.io/api/v2/"
+	defaultSource             = "clair-upstream"
+	defaultKey                = "user_key=9e7da76708fe374d8c10fa752e72989f"
+)
+
+var (
+	supportedEcosystems = []string{"pypi", "maven"}
+)
+
+// Matcher attempts to correlate discovered python packages with reported
+// vulnerabilities.
+type Matcher struct {
+	batchSize          int
+	client             *http.Client
+	ecosystem          string
+	requestConcurrency int
+	url                *url.URL
+	key                string
+	source             string
+}
+
+// Build struct to model CRDA V2 ComponentAnalysis response which
+// delivers Snyk sourced Vulnerability information.
+type Vulnerability struct {
+	FixedIn  []string `json:"fixed_in"`
+	ID       string   `json:"id"`
+	Severity string   `json:"severity"`
+	Title    string   `json:"title"`
+	URL      string   `json:"url"`
+}
+
+type VulnReport struct {
+	Name            string          `json:"name"`
+	Version         string          `json:"version"`
+	Vulnerabilities []Vulnerability `json:"vulnerabilities"`
+}
+
+// Request model.
+type Package struct {
+	Name    string `json:"package"`
+	Version string `json:"version"`
+}
+
+type VulnRequest struct {
+	Ecosystem string    `json:"ecosystem"`
+	Packages  []Package `json:"package_versions"`
+}
+
+// Option controls the configuration of a Matcher.
+type Option func(*Matcher) error
+
+// NewMatcher returns a configured Matcher or reports an error.
+func NewMatcher(ecosystem string, opt ...Option) (*Matcher, error) {
+	if ecosystem == "" {
+		return nil, fmt.Errorf("empty ecosystem")
+	}
+	m := Matcher{ecosystem: ecosystem}
+
+	for _, f := range opt {
+		if err := f(&m); err != nil {
+			return nil, err
+		}
+	}
+
+	if m.url == nil {
+		var err error
+		m.url, err = url.Parse(defaultURL)
+		if err != nil {
+			return nil, err
+		}
+
+	}
+
+	if m.key == "" {
+		m.key = defaultKey
+	} else {
+		m.key = "user_key=" + m.key
+	}
+
+	m.url = &url.URL{
+		Scheme:     m.url.Scheme,
+		Host:       m.url.Host,
+		Path:       defaultEndPoint,
+		ForceQuery: true,
+		RawQuery:   m.key,
```

Comment:
IMO the `m.key` variable shouldn't include the string`"user_key="`, it should be added here:
```go
RawQuery:   "user_key=" + m.key,
```

Reply:
ACK.